src/Catalog/PackageCatalogItemCreator.cs

```diff
@@ -0,0 +1,234 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NuGet.Services.Metadata.Catalog.Helpers;
+using NuGet.Services.Metadata.Catalog.Persistence;
+
+namespace NuGet.Services.Metadata.Catalog
+{
+    public sealed class PackageCatalogItemCreator
+    {
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
+        private readonly IAzureStorage _storage;
+        private readonly ITelemetryService _telemetryService;
+
+        private PackageCatalogItemCreator(
+            HttpClient httpClient,
+            ITelemetryService telemetryService,
+            ILogger logger,
+            IStorage storage)
+        {
+            _httpClient = httpClient;
+            _telemetryService = telemetryService;
+            _logger = logger;
+            _storage = storage as IAzureStorage;
+        }
+
+        public static PackageCatalogItemCreator Create(
+            HttpClient httpClient,
+            ITelemetryService telemetryService,
+            ILogger logger,
+            IStorage storage)
+        {
+            if (httpClient == null)
+            {
+                throw new ArgumentNullException(nameof(httpClient));
+            }
+
+            if (telemetryService == null)
+            {
+                throw new ArgumentNullException(nameof(telemetryService));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            return new PackageCatalogItemCreator(httpClient, telemetryService, logger, storage);
+        }
+
+        public async Task<PackageCatalogItem> CreateAsync(
+            FeedPackageDetails packageItem,
+            DateTime timestamp,
+            CancellationToken cancellationToken)
+        {
+            if (packageItem == null)
+            {
+                throw new ArgumentNullException(nameof(packageItem));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            PackageCatalogItem item = null;
+
+            if (_storage != null)
+            {
+                item = await GetPackageViaStorageAsync(packageItem, cancellationToken);
+            }
+
+            if (item == null)
+            {
+                item = await GetPackageViaHttpAsync(packageItem, timestamp, item, cancellationToken);
+            }
+
+            return item;
+        }
+
+        private async Task<PackageCatalogItem> GetPackageViaStorageAsync(FeedPackageDetails packageItem, CancellationToken cancellationToken)
+        {
+            PackageCatalogItem item = null;
+            var packageName = $"{packageItem.PackageId.ToLowerInvariant()}.{packageItem.PackageVersion.ToLowerInvariant()}.nupkg";
```

Comment:
>{packageItem.PackageId.ToLowerInvariant()}.{packageItem.PackageVersion.ToLowerInvariant()}.nupkg [](start = 32, length = 96)

Not sure, but I suspect that generating package name from id and version is duplicated in a few places in this code base.. consider putting in a shared location.